weaver/quotation/estimation.py

```diff
@@ -0,0 +1,147 @@
+import logging
+import random
+import time
+from typing import TYPE_CHECKING
+
+from pyramid_celery import celery_app as app
+
+from weaver.database import get_db
+from weaver.exceptions import QuoteEstimationError
+from weaver.processes.types import ProcessType
+from weaver.processes.wps_package import get_package_workflow_steps, get_process_location
+from weaver.quotation.status import QuoteStatus
+from weaver.store.base import StoreProcesses, StoreQuotes
+from weaver.utils import fully_qualified_name, get_settings, request_extra, wait_secs
+
+if TYPE_CHECKING:
+    from celery.task import Task
+
+    from weaver.datatype import Process, Quote
+    from weaver.quotation.status import AnyQuoteStatus
+    from weaver.typedefs import AnyUUID
+
+LOGGER = logging.getLogger(__name__)
+
+
+def estimate_process_quote(quote, process):
+    # type: (Quote, Process) -> Quote
+    """
+    Estimate execution price and time for an atomic :term:`Process` operation.
+
+    Employs provided inputs and expected outputs and relevant metadata for the :term:`Process`.
+
+    :param quote: Quote with references to process parameters.
+    :param process: Targeted process for execution.
+    :returns: Updated quote with estimates.
+    """
+
+    # TODO: replace by some fancy ml technique or something?
+    quote.seconds = int(random.uniform(5, 60) * 60 + random.uniform(5, 60))
+    quote.price = float(random.uniform(0, 100) * quote.seconds)
+    quote.currency = "CAD"
+
+    return quote
+
+
+def estimate_workflow_quote(quote, process):
+    # type: (Quote, Process) -> Quote
+    """
+    Loop :term:`Workflow` sub-:term:`Process` steps to get their respective :term:`Quote`.
+    """
+    settings = get_settings()
+    process_url = process.href(settings)
+    quote_steps = []
+    quote_params = []
+    workflow_steps = get_package_workflow_steps(process_url)
+    for step in workflow_steps:
+        # retrieve quote from provider ADES
+        # TODO: data source mapping
+        process_step_url = get_process_location(step["reference"])
+        process_quote_url = "{}/quotations".format(process_step_url)
+
+        # FIXME: how to estimate data transfer if remote process (?)
+        # FIXME: how to produce intermediate process inputs (?)
+        # FIXME: must consider fan-out in case of parallel steps
+        data = {"inputs": [], "outputs": []}
+        resp = request_extra("POST", process_quote_url, json=data, headers={"Prefer": "respond-async"})
+        href = resp.headers.get("Location")
+        status = QuoteStatus.SUBMITTED
+        retry = 0
+        abort = 3
+        while status != QuoteStatus.COMPLETED and abort > 0:
+            wait = wait_secs(retry)
+            retry += 1
+            resp = request_extra("GET", href)
+            if resp.status_code != 200:
+                abort -= 1
+                wait = 5
+            else:
+                body = resp.json()
+                if body.get("status") == QuoteStatus.COMPLETED:
```

Comment:
I see a small risk here : If for some reason, you got a 200 with a quote status != COMPLETED you'll end in an infinite loop. You should add an else clause that decrease abort here too.

Reply:
We could had a global timeout to break out. Otherwise, this is the intended behaviour (ie: loop while still running).

Reply:
Oh, I see. It is missing the update of `status` variable to check against the top `while status != QuoteStatus.COMPLETED`.
So, yes, something to break out when FAILED happens.